File: probe/keys-go/src/fanout.go
package main

// Fan-out: where the runner's view of the graph is FINER than the registry's.
//
// A registry node whose work is really N independent jobs becomes N nodes,
// each with its own key and its own worker slot. Only legal when the jobs are
// genuinely independent — separate outputs, no shared mutable state, each
// failing on its own.
//
// Done here rather than in the generator so the split stays live: adding a
// contract def adds a node without regenerating anything.

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// fanout maps a node id to a splitter. Returning nil leaves the node alone.
// The splitter takes the repo root explicitly: the runner happens to run with
// the root as its cwd, but `go test -C pipeline` does not, and a graph that
// silently comes back without its fan-out parts is the kind of difference
// nothing downstream would notice.
var fanout = map[NodeID]func(root string, n Node) ([]Node, error){
	NContracts: fanContracts,
}

// contracts runs 29 components, already one child process each, serially:
// 46% of the full tier spent on jobs that never touch one another.
func fanContracts(root string, n Node) ([]Node, error) {
	entries, err := os.ReadDir(filepath.Join(root, "tools/contracts/components"))
	if err != nil {
		return nil, err
	}
	var names []string
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".mjs") {
			names = append(names, strings.TrimSuffix(e.Name(), ".mjs"))
		}
	}
	sort.Strings(names)

	// the two broad patterns cover every def, which is what we are narrowing;
	// the harness files they also matched are named explicitly
	var shared []string
	for _, p := range n.Inputs {
		if p != "tools/contracts/**/*.mjs" && p != "tools/contracts/components/**" {
			shared = append(shared, p)
		}
	}

	out := make([]Node, 0, len(names))
	for _, name := range names {
		c := n
		c.ID = NodeID("contracts:" + name)
		c.Run = [][]string{{"./build/pipeline", "contract", name}}
		c.Inputs = append([]string{
			"tools/contracts/components/" + name + ".mjs",
		}, shared...)
		c.Produces = []string{"tools/contracts/out/" + name}
		out = append(out, c)
	}
	return out, nil
}

// expandFanout rewrites the node list, rewiring every `needs` that pointed at
// a split node to point at all of its parts.
func expandFanout(root string, in []Node) ([]Node, error) {
	replaced := map[NodeID][]NodeID{}
	var out []Node
	for _, n := range in {
		split := fanout[n.ID]
		if split == nil {
			out = append(out, n)
			continue
		}
		parts, err := split(root, n)
		if err != nil {
			return nil, err
		}
		ids := make([]NodeID, 0, len(parts))
		for _, p := range parts {
			ids = append(ids, p.ID)
		}
		replaced[n.ID] = ids
		out = append(out, parts...)
	}
	if len(replaced) == 0 {
		return out, nil
	}
	for i := range out {
		var needs []NodeID
		for _, d := range out[i].Needs {
			if parts, ok := replaced[d]; ok {
				needs = append(needs, parts...)
			} else {
				needs = append(needs, d)
			}
		}
		out[i].Needs = needs
	}
	return out, nil
}

File: src/fanout.rs
//! Port of pipeline/fanout.go: where the runner's view of the graph is FINER
//! than the registry's. `contracts` becomes one node per contract def, each
//! with its own key and its own worker slot.

use crate::nodes::Node;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn fan_contracts(root: &Path, n: &Node) -> std::io::Result<Vec<Node>> {
    let entries = fs::read_dir(root.join("tools/contracts/components"))?;
    let mut names: Vec<String> = entries
        .flatten()
        .map(|e| e.file_name().to_string_lossy().into_owned())
        .filter(|f| f.ends_with(".mjs"))
        .map(|f| f.trim_end_matches(".mjs").to_string())
        .collect();
    names.sort();

    // the two broad patterns cover every def, which is what we are narrowing;
    // the harness files they also matched are named explicitly
    let shared: Vec<String> = n
        .inputs
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .filter(|p| *p != "tools/contracts/**/*.mjs" && *p != "tools/contracts/components/**")
        .cloned()
        .collect();

    let mut out = Vec::with_capacity(names.len());
    for name in &names {
        let mut c = n.clone();
        c.id = format!("contracts:{}", name);
        c.run = vec![vec![
            crate::nodes::engine_argv0(),
            "contract".to_string(),
            name.clone(),
        ]];
        let mut inputs = vec![format!("tools/contracts/components/{}.mjs", name)];
        inputs.extend(shared.clone());
        c.inputs = Some(inputs);
        c.produces = Some(vec![format!("tools/contracts/out/{}", name)]);
        out.push(c);
    }
    Ok(out)
}

/// expandFanout rewrites the node list, rewiring every `needs` that pointed at
/// a split node to point at all of its parts.
pub fn expand_fanout(root: &Path, nodes: Vec<Node>) -> Result<Vec<Node>, String> {
    let mut replaced: HashMap<String, Vec<String>> = HashMap::new();
    let mut out: Vec<Node> = Vec::new();
    for n in nodes {
        if n.id == "contracts" {
            let parts =
                fan_contracts(root, &n).map_err(|e| format!("tools/contracts/components: {}", e))?;
            replaced.insert(
                n.id.clone(),
                parts.iter().map(|p| p.id.clone()).collect(),
            );
            out.extend(parts);
            continue;
        }
        out.push(n);
    }
    if replaced.is_empty() {
        return Ok(out);
    }
    for n in out.iter_mut() {
        let mut needs = Vec::with_capacity(n.needs.len());
        for d in &n.needs {
            match replaced.get(d) {
                Some(parts) => needs.extend(parts.iter().cloned()),
                None => needs.push(d.clone()),
            }
        }
        n.needs = needs;
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn contracts_fans_out_and_needs_rewire() {
        let tmp = std::env::temp_dir().join(format!("shadless-rs-fanout-{}", std::process::id()));
        let comp = tmp.join("tools/contracts/components");
        fs::create_dir_all(&comp).unwrap();
        for name in ["dialog.mjs", "button.mjs", "readme.txt"] {
            fs::write(comp.join(name), "").unwrap();
        }
        let list = expand_fanout(&tmp, crate::nodes::all()).unwrap();
        // 40 authored nodes + 2 shards replacing contracts
        assert_eq!(list.len(), 42);
        let shards: Vec<&Node> = list.iter().filter(|n| n.id.starts_with("contracts:")).collect();
        assert_eq!(
            shards.iter().map(|n| n.id.as_str()).collect::<Vec<_>>(),
            vec!["contracts:button", "contracts:dialog"]
        );
        // shard run/inputs/produces — argv0 must be whatever the presented
        // graph runs (Go path under SHADLESS_GRAPH=go-mirror, else __self__)
        assert_eq!(
            shards[0].run,
            vec![vec![
                crate::nodes::engine_argv0(),
                "contract".to_string(),
                "button".to_string()
            ]]
        );
        assert_eq!(
            shards[0].produces.as_deref(),
            Some(&["tools/contracts/out/button".to_string()][..])
        );
        assert_eq!(
            shards[0].inputs.as_deref().unwrap().first(),
            Some(&"tools/contracts/components/button.mjs".to_string())
        );
        // the broad def-glob is narrowed out of the shared inputs
        assert!(!shards[0]
            .inputs
            .as_deref()
            .unwrap()
            .contains(&"tools/contracts/components/**".to_string()));
        // style-parity's needs rewired to every shard, in sorted order
        let sp = list.iter().find(|n| n.id == "style-parity").unwrap();
        assert_eq!(
            sp.needs,
            vec![
                "contracts:button".to_string(),
                "contracts:dialog".to_string(),
                "oracle-css".to_string(),
                "build-js".to_string()
            ]
        );
        let _ = fs::remove_dir_all(&tmp);
    }
}
